drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1303 @@
+			nrfx_i2s_buffers_t const *p_released,
+			u32_t status, nrfx_i2s_buffers_t *p_new_buffers)
+{
+	struct channel_str *ch_rx = &i2s->channel_rx;
+	int ret;
+
+	if (p_released->p_rx_buffer != NULL && next_buffers_needed(status)) {
+		/* Content of received buffer is valuable.
+		 * If 'EVENT_STOPPED' generated 'then next_buffers_needed()'
+		 * returns false - function 'channel_rx_store_data()' won't
+		 * execute because buffer didn't fill.
+		 */
+		ret = channel_rx_store_data(i2s,
+					   (u32_t **)&p_released->p_rx_buffer);
+		if (ret < 0) {
+			return;
+		}
+	}
+	if (ch_rx->current_state == I2S_STATE_STOPPING) {
+		/* finishing rx transfer caused by user trigger command*/
+		enum i2s_trigger_cmd ch_cmd = ch_rx->last_trigger_cmd;
+
+		if (next_buffers_needed(status)) {
+			switch (ch_cmd) {
+			case I2S_TRIGGER_DROP:
+				channel_rx_mem_clear(i2s);
+				break;
+			case I2S_TRIGGER_DRAIN:
+			case I2S_TRIGGER_STOP:
+				break;
+			default:
+				LOG_ERROR("RX Callback command unknown");
+				channel_change_state(ch_rx, I2S_STATE_ERROR);
+				return;
+			}
+		} else {
+			if (p_released->p_rx_buffer) {
+				ret = channel_rx_store_data(i2s,
+					(u32_t **)&p_released->p_rx_buffer);
+				if (ret < 0) {
+					return;
+				}
+			}
+			int ret = channel_change_state(ch_rx,
+					I2S_STATE_READY);
+			if (ret < 0) {
+				channel_change_state(ch_rx, I2S_STATE_ERROR);
+				return;
+			}
+		}
+		i2s->buffers.p_rx_buffer = NULL;
+		return;
+	} else if (ch_rx->current_state == I2S_STATE_RUNNING &&
+		   interface_get_state(i2s) == I2S_IF_NEEDS_RESTART) {
+		return;
+	} else if (ch_rx->current_state == I2S_STATE_ERROR) {
+		if (p_released->p_rx_buffer != NULL) {
+			k_mem_slab_free(ch_rx->mem_slab,
+					(void **)&p_released->p_rx_buffer);
+		}
+		return;
+	}
+	if (next_buffers_needed(status)) {
+		ret = k_mem_slab_alloc(ch_rx->mem_slab,
+				       (void **)&p_new_buffers->p_rx_buffer,
+				       K_NO_WAIT);
+		if (ret < 0) {
+			/*overrun error occurred*/
+			interface_error_service(i2s, "RX overrun error");
+			channel_change_state(ch_rx, I2S_STATE_ERROR);
+			return;
+		}
+	}
+}
+
+static int channel_rx_store_data(struct i2s_nrfx_data *i2s, u32_t **buf)
+{
+	struct channel_str *ch_rx = &i2s->channel_rx;
+	int ret;
+
+	assert(i2s != NULL);
+	ret = queue_add(&ch_rx->mem_block_queue, *buf);
+	if (ret < 0) {
+		return ret;
+	}
+	k_sem_give(&ch_rx->sem);
+
+	return 0;
+}
+
+static void isr(void *arg)
+{
+	struct i2s_nrfx_data *i2s = get_interface();
+
+	/* 'nrfx_i2s_irq_handler()' calls 'interface_handler()' which in turn
+	 * can call:
+	 *  - 'channel_tx_callback()' when tx channel is running
+	 *  - 'channel_tx_callback()' when rx channel is running
+	 */
+	nrfx_i2s_irq_handler();
+	/* 'interface_handler()' can be called twice per one 'isr()'
+	 * execution. Every 'interface_handler()' call can modify state of
+	 * interface (i2s->state). After all it is checked and started again
+	 * if needed.
```

Comment:
But this does not justify the placing of this code here, as the interface's state could be checked at the end of `interface_handler`. It rather seems that the actual cause is that currently it is not possible to call `nrfx_i2s_start` from the handler function passed to the nrfx_i2s driver (`interface_handler` in this case). But this would mean that this piece of code is a workaround for the current limitation of the nrfx_i2s driver, and it should be described as such. And after the driver is corrected in this regard, this workaround should be removed.